crates/handler/src/pre_execution.rs

```diff
@@ -106,73 +125,84 @@ pub fn validate_account_nonce_and_code(
+    balance: U256,
+    tx: impl Transaction,
+    block: impl Block,
+    cfg: impl Cfg,
+) -> Result<U256, InvalidTransaction> {
+    let basefee = block.basefee() as u128;
+    let blob_price = block.blob_gasprice().unwrap_or_default();
+    let is_balance_check_disabled = cfg.is_balance_check_disabled();
+
+    if !is_balance_check_disabled {
+        tx.has_enough_balance(balance)?;
     }
 
     let effective_balance_spending = tx
         .effective_balance_spending(basefee, blob_price)
         .expect("effective balance is always smaller than max balance so it can't overflow");
 
-    // subtracting max balance spending with value that is going to be deducted later in the call.
     let gas_balance_spending = effective_balance_spending - tx.value();
 
-    let mut new_balance = caller_account
-        .info
-        .balance
-        .saturating_sub(gas_balance_spending);
+    // new balance
+    let mut new_balance = balance.saturating_sub(gas_balance_spending);
 
     if is_balance_check_disabled {
         // Make sure the caller's balance is at least the value of the transaction.
         new_balance = new_balance.max(tx.value());
     }
 
-    let old_balance = caller_account.info.balance;
+    Ok(new_balance)
+}
+
+/// Make changes to the caller account.
+///
+/// Returns the old balance.
+#[inline]
+pub fn caller_touch_and_change(
+    caller_account: &mut Account,
+    new_balance: U256,
+    is_call: bool,
+) -> U256 {
     // Touch account so we know it is changed.
     caller_account.mark_touch();
-    caller_account.info.balance = new_balance;
-
+    let old_balance = core::mem::replace(&mut caller_account.info.balance, new_balance);
     // Bump the nonce for calls. Nonce for CREATE will be bumped in `make_create_frame`.
-    if tx.kind().is_call() {
+    if is_call {
         // Nonce is already checked
         caller_account.info.nonce = caller_account.info.nonce.saturating_add(1);
     }
+    old_balance
+}
+
+/// Validates caller state and deducts transaction costs from the caller's balance.
+#[inline]
+pub fn validate_against_state_and_deduct_caller<
+    CTX: ContextTr,
+    ERROR: From<InvalidTransaction> + From<<CTX::Db as Database>::Error>,
+>(
+    context: &mut CTX,
+) -> Result<(), ERROR> {
+    let (tx, block, cfg, journal) = context.tx_block_cfg_journal_mut();
 
+    // Load caller's account.
+    let caller_account = journal.load_account_code(tx.caller())?.data;
+
+    // validate account nonce and code
+    validate_account_nonce_and_code_with_components(&mut caller_account.info, tx, cfg)?;
+
+    let new_balance =
+        deduct_caller_balance_with_components(caller_account.info.balance, tx, block, cfg)?;
+
+    // make changes to the account
+    let old_balance = caller_touch_and_change(caller_account, new_balance, tx.kind().is_call());
+
+    // journal the change
     journal.caller_accounting_journal_entry(tx.caller(), old_balance, tx.kind().is_call());
+
```

Comment:
@klkvr @mattsse can you take a look? This function is a lot simpler than before as few things got abstracted with functions